backend/src/app.ts

```diff
@@ -115,7 +122,54 @@ app.get('/reviews', async (req, res) => {
     );
     res.status(200).send(JSON.stringify(resultsWithType));
   } catch (err) {
-    res.status(400).send(err);
+    console.error(err);
+    res.status(400).send('Error');
+  }
+});
+
+app.post('/add-like', authenticate, async (req, res) => {
+  try {
+    if (!req.user) throw new Error('not authenticated');
+    const { uid } = req.user;
+    const { reviewId } = req.body;
+    if (!reviewId) throw new Error('must specify review id');
+    const likesRef = likesCollection.doc(uid);
+    const reviewRef = reviewCollection.doc(reviewId);
+    await db.runTransaction(async (t) => {
+      const likesDoc = await t.get(likesRef);
+      const result = likesDoc.get(reviewId);
+      if (!result) {
+        t.set(likesRef, { [reviewId]: true }, { merge: true });
+        t.update(reviewRef, { likes: FieldValue.increment(1) });
+      }
+    });
+    res.status(200).send(JSON.stringify({ result: 'Success' }));
+  } catch (err) {
+    console.error(err);
+    res.status(400).send('Error');
+  }
+});
+
+app.post('/remove-like', authenticate, async (req, res) => {
+  try {
+    if (!req.user) throw new Error('not authenticated');
+    const { uid } = req.user;
+    const { reviewId } = req.body;
+    if (!reviewId) throw new Error('must specify review id');
+    const likesRef = likesCollection.doc(uid);
+    const reviewRef = reviewCollection.doc(reviewId);
+    await db.runTransaction(async (t) => {
+      const likesDoc = await t.get(likesRef);
+      const result = likesDoc.get(reviewId);
+      if (result) {
+        likesRef.set({ [reviewId]: FieldValue.delete() }, { merge: true });
+        reviewRef.update({ likes: FieldValue.increment(-1) });
+      }
+    });
+    res.status(200).send(JSON.stringify({ result: 'Success' }));
+  } catch (err) {
+    console.error(err);
+    res.status(400).send('Error');
   }
 });
```

Comment:
consider putting the body of the functions here in a helper function with maybe a boolean on which operation you're doing in order to deduplicate the code